### ISSUE TITLE
Add error handling to VimwikiSearch

### ISSUE THREAD

**Issue by glappen** [vimwiki/vimwiki#420]
Hello, trying to get this setup the way I want and I have made 2 different wiki's:

```
let g:vimwiki_list = [{'path': '/home/greg/Nextcloud/vimwiki/Personal'}, {'path': '/home/greg/Nextcloud/vimwiki/LapComInc'}]
```

When I open a wiki and run :VimwikiSearch, I get this error:

```
"reg/Nextcloud/vimwiki/LapComInc/**/*.wiki" [New DIRECTORY]                                                                                                    
Cannot open file "reg/Nextcloud/vimwiki/LapComInc/**/*.wiki"                                                                                                   
E480: No match: home                                                                                                                                           
Press ENTER or type command to continue
```

The wikis seem to function normally otherwise.  Can anyway tell me where I went wrong here?

Thanks!!!

**Comment by glappen** [comment]
I guess this was a really obscure error...is that the correct way to setup two different wikis at least?  Why is VimwikiSearch getting the path to the wiki files wrong?

**Comment by Nudin** [comment]
Hi,
did you specify a search-string/pattern? Like `:VimwikiSearch foo`? If not – the error is not surprising: Vimwiki does not (jet) check is an argument is given – VimwikiSearch is simply bound to
`lvimgrep <args> yourpath/**/*.wiki`. If you don't give any argument, you get:
```
lvimgrep /home/greg/Nextcloud/vimwiki/LapComInc
```
So vim thinks `/home/g` is your search-pattern and searches for this in the path `reg/Nextcloud/vimwiki/LapComInc`.

One should probably handle the case when the user does not give an argument gracefully…

**Comment by ranebrown** [comment]
The command should be changed to call a function with error handling built in.

**Comment by ratfactor** [comment]
I'd like to contribute to VimWiki.  If no one else has started on this issue, it seems like a nice one for me to tackle.

**Comment by ranebrown** [comment]
That would be great! Make sure to take a look at the test framework and include new tests as part of your PR. Let me know if you have any questions on that front.

**Comment by ratfactor** [comment]
I have the fix completed.  Starting on tests.  I was very happy to see (and run) the test framework in the dev branch.  Good things are happening in this project.

**Comment by ratfactor** [comment]
While I'm in here, I've noticed two other improvements to usability for VimwikiSearch:

### Unenclosed search patterns

If the user types a search pattern containing spaces without `/enclosing/` it, they'll get an error very similar to the original cause of this issue:

```
:VimwikiSearch foo bar
"bar" [New File]
Cannot open file "bar"
```

The same search with an enclosed pattern works as expected:

```
:VimwikiSearch /foo bar/
```


To solve this, I'm going to enclose any search pattern not enclosed in "non-ID" characters (as per `:help vimgrep` and `:help isdent`) in a pair of '/'s.  I believe this will be the least surprising behavior for most users.

In fact, if the search hasn't been explicitly `/enclosed/` by the user, I'm thinking we should do a "very nomagic" (`:help magic`) search to avoid further surprises:

```
lvimgrep /\V<user search>/
```

Advanced users can still enclose the pattern themselves and get complete control over the search.

### No matches looks like an error

When the VImwikiSearch  results in no matches, Vim (`:lvimgrep`) returns an error:

```
E480: No match: does_not_exist
```

The problem is twofold:
1. It looks like an error (red background in my color scheme) and includes the Vim error number (which I've found difficult to lookup - it's not clear what this error really *means* other than what it says: "No match").
2. Error messages like these disappear after a short time, so if the user looks away, it won't be clear what happened.

To improve this, I'm going to enclose the `lvimgrep` in a `try...catch` and show a friendlier message:

```
echomsg "VimwikiSearch: No match found."
```

The only downside I can imagine is that someone may have written a script to use `VimwikiSearch` and is depending on the `E480` error.   Thoughts?

**Comment by ranebrown** [comment]
I think both of those changes make sense and would make the command more user friendly.

**Comment by ratfactor** [comment]
Okay, I've pushed a commit to a feature branch on my fork here:

https://github.com/ratfactor/vimwiki/tree/issue-420

(I'm a GitHub noob, so let me know if this looks right so far.)

However, I have _not_ made a pull request for one reason:  my Vader tests don't work in the automated environment.  I was afraid of that.  Testing `lvimgrep` in Vader turned out to be challenging, but I did get a series of tests to run reliably...manually.

The basic issue is: any "error" (even something as innocent as a "no match" for a search, and even in a try/catch block) will cause all subsequent `lvimgrep` tests to fail.  That, or something like it, is happening in the automated environment and preventing the tests from succeeding. :disappointed: 

So, either someone with deeper Vim mastery can help figure out how to get the tests to pass, *or* perhaps we just move the search tests to something like a `test/manual` directory.

Other than that, Vader works great and the tests helped me verify that VWS was still working as I made changes.

**Comment by ranebrown** [comment]
@ratfactor I don't have time to look into the Vader tests right now to see what might be causing the tests to fail after getting an error message. I would suggest two things:
1. Create an [issue](https://github.com/junegunn/vader.vim/issues) in the Vader project and see if you can get an answer there.
2. See if it would be possible to write a bash script to automate the remaining tests that you ran manually.

If neither of those options pan out go ahead and create the manual directory with instructions you used to run the tests. Once your PR is merged we can create a new issue to track converting those to some sort of automated test.

How does that sound?

**Comment by ratfactor** [comment]
@ranebrown Sounds good.  I've done Suggestion One and made a test case and written a plea for help with the Vader folk:

https://github.com/junegunn/vader.vim/issues/199

The tests I've written do run fine from the command line, so Suggestion Two should be a fine fallback if I can't get past this corner case testing as part of the rest of the suite.
